### PR TITLE
Added : Assignment Close Immediately 

### LINKED ISSUE #1766: Assignment Close Immediately
Currently assignments close at a deadline. However, it will be a nice for the professor to have the ability to close the assignment immediately. This will be helpful in a test like environment, where the official time is 1 hour, but the professor has set deadline as 1.5 hours (for margin). But once he is convinced everyone is done, he can close it manually himself. 

I would like a button which says "Close now"

**Are you working on this? (Yes/No)**
No

### PR DESCRIPTION
**Pull request by pulkit-30**
Fixes #1766 

#### Describe the changes you have made in this PR -

Added a ` close ` button in Assignment that closes the assignment immediately.

### Screenshots of the changes (If any) -

https://user-images.githubusercontent.com/76155456/148927346-f870ee3d-858a-4c59-8de9-eeb16079025b.mov

Please provide me with appropriate feedback so that I can continue working on this

### CODE REVIEW

File: app/policies/assignment_policy.rb [hunk 1]
```diff
@@ -33,6 +33,12 @@ def reopen?
     true
   end
 
+  def close?
+    raise CustomAuthError, "Project is already closed" if assignment.status == "closed"
+
+    true
```
Comment (tachyons) [inline app/policies/assignment_policy.rb:39]: This allows all users to close assignment, which is a big security hole

File: config/routes.rb [hunk 2]
```diff
@@ -45,6 +45,7 @@
 
   scope "/groups" do
     get "/:group_id/assignments/:id/reopen", to: "assignments#reopen", as: "reopen_group_assignment"
+    get "/:group_id/assignments/:id/close", to: "assignments#close", as: "close_group_assignment"
```
Comment (tachyons) [inline config/routes.rb:48]: Make this put request

File: app/controllers/assignments_controller.rb [hunk 3]
```diff
@@ -59,6 +59,16 @@ def reopen
     redirect_to edit_group_assignment_path(@group, @assignment)
   end
 
+  #close assignment
+  def close
```
Comment (tachyons) [inline app/controllers/assignments_controller.rb:63]: Add this logic in api as well

File: app/policies/assignment_policy.rb (hunk 1 above)
Comment (pulkit-30) [reply app/policies/assignment_policy.rb:39]: May you please help me out, how to solve this !
as far as I can see,

other members have no option to close assignment  
Screenshot 2022-01-11 at 5 25 20 PM

Thanks in advance 🤗

File: app/controllers/assignments_controller.rb (hunk 3 above)
Comment (pulkit-30) [reply app/controllers/assignments_controller.rb:63]: Added please review!

File: app/policies/assignment_policy.rb (hunk 1 above)
Comment (tachyons) [reply app/policies/assignment_policy.rb:39]: You are just hiding it in the ui, an attacker can goto the link even if it is not enabled on the ui

File: app/controllers/api/v1/assignments_controller.rb [hunk 4]
```diff
@@ -62,6 +62,18 @@ def reopen
     end
   end
 
+  # PUT /api/v1/assignments/:id/close
+  def close
```
Comment (tachyons) [inline app/controllers/api/v1/assignments_controller.rb:66]: - Missing i18n
- Missing authorisation

**Comment by pulkit-30** [comment]
@tachyons , please review

**Comment by tachyons** [comment]
Awesome 💯

**Pull request merged**